encoding/descaling.md

```diff
@@ -0,0 +1,356 @@
+
+This picture is dark.
+It also has some effects over it.
+
+![Manaria Friends Ep01 frame 2](images/descale_manaria02.png)
+
+This picture is also very dark,
+and has even more effects over it.
+
+![Manaria Friends Ep01 frame 3](images/descale_manaria03.png)
+
+Heavy dynamic grain will almost always give bad results.
+
+![Manaria Friends Ep01 frame 4](images/descale_manaria04.png)
+
+This is a nice frame to use as reference.
+The background is a bit blurry,
+but it isn't full of effects
+and is fairly bright.
+The lineart is very clear.
+
+We will now make use of the getnative.py script
+to figure out what resolution
+this anime was produced at.
+Run the following in your terminal:
+
+```bash
+$ python getnative.py "Manaria Friends Ep01 frame 4.png"
+```
+
+It should show the following:
+
+```
+Using imwri as source filter
+501/501
+Kernel: bicubic AR: 1.78 B: 0.33 C: 0.33
+Native resolution(s) (best guess): 878p
+done in 18.39s
+```
+
+If you look in the directory
+that you ran the script in,
+you will find a new folder
+called "getnative".
+You can find the following graph
+in there as well:
+
+![Manaria_Friends-ep1_04_graph](images/descale_graph.png)
+
+You should be looking for a low relative error.
+In this case it very clearly points to 878p.
+
+
+## Descaling
+
+Now it's time to actually start descaling.
+Open up your Vapoursynth editor of choice,
+and import the clip:
+
+```Py
+src = lvf.src(r"BDMV/[BDMV][190302][マナリアフレンズ I]/BD/BDMV/STREAM/00007.m2ts")
+```
+
+The next issue is figuring out
+what was used to [upsample](resampling.md#upsampling) the show.
+By default,
+getnative.py checks with Mitchell-Netravali
+(bicubic b=1/3, c=1/3).
+However, it might have also been upscaled
+using other kernels,
+like Spline or Bilinear.
+The best way to figure out
+what is used is to simply try out
+a bunch of and use your eyes.
+Check for common scaling-related artifacting,
+like haloing,
+ringing,
+aliasing,
+etc.
+
+For bicubic it is important
+to keep in mind that
+you will typically find that
+the values match the following mathmatical expressions:
+
+`b + 2c = 1` or `b + c = 1`
+
+It could also just be 0 or 1.
+
+Whilst this isn't a 100% guarantee,
+this is the most common approach
+to resampling using bicubic,
+so it's worth keeping in mind.
+
+Here's an example of the previous frame
+when descaled using various kernels and settings
+(note that descale requires either GrayS, RGBS, or YUV444PS.
+I'll be using `split` and `join` from `kagefunc` to split the planes
+and then join them again in this example,
+and `get_w` from `vsutil` to calculate the width)[^1]:
+
+[Comparison between frames][manaria_compare]
+
+```Py
+import vapoursynth as vs
+import vsutil
+import kagefunc as kgf
+import fvsfunc as fvf
+core = vs.core
+
+src = core.lsmas.LWLibavSource("BDMV/[BDMV][190302][マナリアフレンズ I]/BD/BDMV/STREAM/00007.m2ts")
+src = fvf.Depth(src, 32)
+
+Y, U, V = kgf.split(src)
+width = vsutil.get_w(878)
+height = 878
```

Comment:
Reversing the lines shows that getnative outputs height and you work around that:

```Py
height = 878
width = vsutil.get_w(height)
```

Reply:
Changed